### ISSUE TITLE
Segmentation fault in wazuh-logcollector

### ISSUE THREAD

**Issue by legioner0** [wazuh/wazuh#23742]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.7.4-1 | wazuh-logcollector | Agent | Wazuh Repo | Ubuntu 22.04.4 LTS |

`wazuh-logcollector` crashes at random moments - sometimes work for days without any problem, sometimes - crashes every 30 minutes. I tried setting `logcollector.debug=2`, but it didn't output anything useful when it crashed.

**ossec.conf**
```
<ossec_config>
  <client>
    <server>
      <address>skipped</address>
      <port>1514</port>
      <protocol>tcp</protocol>
    </server>
    <config-profile>ubuntu, ubuntu22, ubuntu22.04</config-profile>
    <notify_time>10</notify_time>
    <time-reconnect>60</time-reconnect>
    <auto_restart>yes</auto_restart>
    <crypto_method>aes</crypto_method>
    <enrollment>
      <enabled>yes</enabled>
      <manager_address>skipped</manager_address>
      <port>1515</port>
      <agent_name>skipped</agent_name>
      <groups>skipped</groups>
      <server_ca_path>/etc/wazuh-data/root-ca.crt</server_ca_path>
      <agent_certificate_path>/etc/wazuh-data/agent.crt</agent_certificate_path>
      <agent_key_path>/etc/wazuh-data/agent.key</agent_key_path>
      <authorization_pass_path>etc/authd.pass</authorization_pass_path>
    </enrollment>
  </client>

  <client_buffer>
    <!-- Agent buffer options -->
    <disabled>no</disabled>
    <queue_size>5000</queue_size>
    <events_per_second>500</events_per_second>
  </client_buffer>

  <!-- Policy monitoring -->
  <rootcheck>
    <disabled>no</disabled>
    <check_files>yes</check_files>
    <check_trojans>yes</check_trojans>
    <check_dev>yes</check_dev>
    <check_sys>yes</check_sys>
    <check_pids>yes</check_pids>
    <check_ports>yes</check_ports>
    <check_if>yes</check_if>

    <!-- Frequency that rootcheck is executed - every 12 hours -->
    <frequency>43200</frequency>

    <rootkit_files>etc/shared/rootkit_files.txt</rootkit_files>
    <rootkit_trojans>etc/shared/rootkit_trojans.txt</rootkit_trojans>

    <skip_nfs>yes</skip_nfs>
  </rootcheck>

  <wodle name="cis-cat">
    <disabled>yes</disabled>
    <timeout>1800</timeout>
    <interval>1d</interval>
    <scan-on-start>yes</scan-on-start>

    <java_path>wodles/java</java_path>
    <ciscat_path>wodles/ciscat</ciscat_path>
  </wodle>

  <!-- Osquery integration -->
  <wodle name="osquery">
    <disabled>yes</disabled>
    <run_daemon>yes</run_daemon>
    <log_path>/var/log/osquery/osqueryd.results.log</log_path>
    <config_path>/etc/osquery/osquery.conf</config_path>
    <add_labels>yes</add_labels>
  </wodle>

  <!-- System inventory -->
  <wodle name="syscollector">
    <disabled>no</disabled>
    <interval>1h</interval>
    <scan_on_start>yes</scan_on_start>
    <hardware>yes</hardware>
    <os>yes</os>
    <network>yes</network>
    <packages>yes</packages>
    <ports all="no">yes</ports>
    <processes>yes</processes>

    <!-- Database synchronization settings -->
    <synchronization>
      <max_eps>10</max_eps>
    </synchronization>
  </wodle>

  <sca>
    <enabled>yes</enabled>
    <scan_on_start>yes</scan_on_start>
    <interval>12h</interval>
    <skip_nfs>yes</skip_nfs>
  </sca>

  <!-- File integrity monitoring -->
  <syscheck>
    <disabled>no</disabled>

    <!-- Frequency that syscheck is executed default every 12 hours -->
    <frequency>43200</frequency>

    <scan_on_start>yes</scan_on_start>

    <!-- Directories to check  (perform all possible verifications) -->
    <directories>/etc,/usr/bin,/usr/sbin</directories>
    <directories>/bin,/sbin,/boot</directories>

    <!-- Files/directories to ignore -->
    <ignore>/etc/mtab</ignore>
    <ignore>/etc/hosts.deny</ignore>
    <ignore>/etc/mail/statistics</ignore>
    <ignore>/etc/random-seed</ignore>
    <ignore>/etc/random.seed</ignore>
    <ignore>/etc/adjtime</ignore>
    <ignore>/etc/httpd/logs</ignore>
    <ignore>/etc/utmpx</ignore>
    <ignore>/etc/wtmpx</ignore>
    <ignore>/etc/cups/certs</ignore>
    <ignore>/etc/dumpdates</ignore>
    <ignore>/etc/svc/volatile</ignore>

    <!-- File types to ignore -->
    <ignore type="sregex">.log$|.swp$</ignore>

    <!-- Check the file, but never compute the diff -->
    <nodiff>/etc/ssl/private.key</nodiff>

    <skip_nfs>yes</skip_nfs>
    <skip_dev>yes</skip_dev>
    <skip_proc>yes</skip_proc>
    <skip_sys>yes</skip_sys>

    <!-- Nice value for Syscheck process -->
    <process_priority>10</process_priority>

    <!-- Maximum output throughput -->
    <max_eps>50</max_eps>

    <!-- Database synchronization settings -->
    <synchronization>
      <enabled>yes</enabled>
      <interval>5m</interval>
      <max_eps>10</max_eps>
    </synchronization>
  </syscheck>

  <!-- Log analysis -->
  <localfile>
    <log_format>command</log_format>
    <command>df -P</command>
    <frequency>360</frequency>
  </localfile>

  <localfile>
    <log_format>full_command</log_format>
    <command>netstat -tulpn | sed 's/\([[:alnum:]]\+\)\ \+[[:digit:]]\+\ \+[[:digit:]]\+\ \+\(.*\):\([[:digit:]]*\)\ \+\([0-9\.\:\*]\+\).\+\ \([[:digit:]]*\/[[:alnum:]\-]*\).*/\1 \2 == \3 == \4 \5/' | sort -k 4 -g | sed 's/ == \(.*\) ==/:\1/' | sed 1,2d</command>
    <alias>netstat listening ports</alias>
    <frequency>360</frequency>
  </localfile>

  <localfile>
    <log_format>full_command</log_format>
    <command>last -n 20</command>
    <frequency>360</frequency>
  </localfile>

  <!-- Active response -->
  <active-response>
    <disabled>no</disabled>
    <ca_store>etc/wpk_root.pem</ca_store>
    <ca_verification>yes</ca_verification>
  </active-response>

  <!-- Choose between "plain", "json", or "plain,json" for the format of internal logs -->
  <logging>
    <log_format>plain</log_format>
  </logging>

</ossec_config>

<ossec_config>
  <localfile>
    <log_format>syslog</log_format>
    <location>/var/ossec/logs/active-responses.log</location>
  </localfile>

  <localfile>
    <log_format>syslog</log_format>
    <location>/var/log/auth.log</location>
  </localfile>

  <localfile>
    <log_format>syslog</log_format>
    <location>/var/log/syslog</location>
  </localfile>

  <localfile>
    <log_format>syslog</log_format>
    <location>/var/log/dpkg.log</location>
  </localfile>

  <localfile>
    <log_format>syslog</log_format>
    <location>/var/log/kern.log</location>
  </localfile>

  <wodle name="docker-listener">
    <disabled>no</disabled>
  </wodle>
  <localfile>
    <log_format>multi-line-regex</log_format>
    <location>/var/log/docker/*</location>
    <out_format>$(timestamp) $(location): $(log)</out_format>
    <!--
         Multiline messages regex. Supports:
         Ruby: "[UFWID], <year>-<month>-<day>T<hour>:<minute>:<second>.<microsecond> "
         Java: "<year>-<month>-<day> <hour>:<minute>:<second>,<millisecond> ["
    -->
    <multiline_regex match="start">((?:[UFWID],\s+\[\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{6}\s+)|(?:\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2}:\d{2},\d{3}\s+\[))</multiline_regex>
  </localfile>
  <labels>
    <label key="instance-id">skipped</label>
  </labels>
</ossec_config>
```

**coerdump info**
```
           PID: 109038 (wazuh-logcollec)
           UID: 0 (root)
           GID: 120 (wazuh)
        Signal: 11 (SEGV)
     Timestamp: Wed 2024-05-29 00:06:19 MSK (9h ago)
  Command Line: /var/ossec/bin/wazuh-logcollector
    Executable: /var/ossec/bin/wazuh-logcollector
 Control Group: /system.slice/wazuh-logcollector-watchdog.service
          Unit: wazuh-logcollector-watchdog.service
         Slice: system.slice
       Boot ID: 87de2224426342fc99957fc8b834d58f
    Machine ID: f6cd92d902794443ac1b94f0bf710af0
      Hostname: skipped
       Storage: /var/lib/systemd/coredump/core.wazuh-logcollec.0.87de2224426342fc99957fc8b834d58f.109038.1716930379000000.zst (present)
     Disk Size: 292.5K
       Message: Process 109038 (wazuh-logcollec) of user 0 dumped core.

                Found module /var/ossec/bin/wazuh-logcollector without build-id.
                Found module /var/ossec/lib/libwazuhext.so without build-id.
                Found module linux-vdso.so.1 with build-id: d0925f56dcc44129a9fbd6897ae9a5a5aab11f7b
                Found module ld-linux-x86-64.so.2 with build-id: 246ac0d8deba5a40c63e9a1a87d4d779d8eb589f
                Found module libc.so.6 with build-id: 962015aa9d133c6cbcfb31ec300596d7f44d3348
                Found module libpthread.so.0 with build-id: 81f46d553e2f7c999e43c3eede73a822bc8d5d93
                Found module libm.so.6 with build-id: b55ba6d8b5954b479fab2a69a54b9f56451fbee0
                Found module libdl.so.2 with build-id: 9f5091f9afcafa6a303ac0d491bce7e58eaf733e
                Found module librt.so.1 with build-id: e90642a19b7c75be941631490081850c92677665
                Stack trace of thread 109051:
                #0  0x00007fefc8786f15 n/a (libc.so.6 + 0x1aff15)
                #1  0x0000000000410b9a n/a (/var/ossec/bin/wazuh-logcollector + 0x10b9a)
```

I have coredump file, but can't share it here, because it contains log lines. I can share it private by email.

**Comment by juliancnn** [comment]
Hi @legioner0,

Reviewing your configuration I have seen the following block:

```xml
  <localfile>
    <log_format>multi-line-regex</log_format>
    <location>/var/log/docker/*</location>
    <out_format>$(timestamp) $(location): $(log)</out_format>
    <multiline_regex match="start">((?:[UFWID],\s+\[\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{6}\s+)|(?:\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2}:\d{2},\d{3}\s+\[))</multiline_regex>
  </localfile>
```
Currently there is a known bug when using wildcards in multiline regex that causes a crash. This bug has already been fixed in PR: [Fix segmentation fault in multiline-regex configuration #23468 ](https://github.com/wazuh/wazuh/pull/23468) and will be released in the next wazuh patch. 

Temporarily, avoid using wildcards (`*`) when the log_format is multi-line-regex to prevent this problem.

I hope this is helpful

**Comment by legioner0** [comment]
Ok, let's assume this issue is duplicate. I have a workaround with tiny watchdog script and can wait for next release.

**Comment by JcabreraC** [comment]
As previously mentioned, this bug has been fixed in Wazuh version 4.7.5, which has already been released. After updating to this version, you should be able to use this feature without any issues.

You can review the applied fix here: [GitHub Pull Request #23468](https://github.com/wazuh/wazuh/pull/23468)
